services/QuillLMS/app/services/diagnostic_insights/data_integrity/verify_by_email.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  module DataIntegrity
+    class VerifyByEmail < ApplicationService
+      attr_reader :email_addresses
+
+      def initialize(email_addresses)
+        @email_addresses = email_addresses
+      end
+
+      def run
+        compact_all_true(users.to_h do |user|
+          [user.email, compact_all_true(verify_user_data_integrity(user))]
+        end)
+      end
+
+      private def users = @users ||= User.where(email: email_addresses)
+
+      # Recursively collapses hashes in which all values are `true` into the simple value `true`.
+      # This allows us to report back cases where data doesn't match, but to avoid having to parse cases where data matches
+      private def compact_all_true(data)
+        recursed_data = data.transform_values do |value|
```

Comment:
Are we worried about this recursive function eating up memory? Maybe there's a way to do this without iteration, like filtered out and not calculating the data for matches?